accepted/2023/nuget-config-apikeys-unencrypted.md

```diff
@@ -0,0 +1,115 @@
+
+The `<packageSources>` section also uses the `<add key="string" value="string">` syntax that most other parts of `nuget.config` use, but it also supports the `protocolVersion="int"` attribute.
+The approach it takes to add `protocolVersion` can be copied to make the `encrypted` attribute work.
+
+## Drawbacks
+
+<!-- Why should we not do this? -->
+
+### Backwards compatibility
+
+There are risks in changing the `nuget.config` schema, particularly around compatibility with older versions of NuGet.
+This includes NuGet.exe, the `dotnet` CLI, Visual Studio, Visual Studio for Mac, and possibly 3rd party IDEs such as JetBrains Rider.
+This is compounded by the fact that the `dotnet` CLI supports `global.json` to pin to older versions of the `dotnet` CLI (and therefore SDK) within a directory.
+
+Problems with the above proposal that extends the existing schema with an additional `encrypted="true/false"` property.
+
+1. Reading nuget.config from older versions of NuGet.
+
+It needs to be verified if older versions of NuGet will fail to load the file if it contains this unexpected attribute on the `<add>` element.
+However, even if it doesn't, since older versions of NuGet expect the value to be always encrypted, it will try to decrypt with DPAPI, and there's a very low chance that the value will decrypt, so it's likely that it will throw an exception saying that the value isn't valid.
+Customers in the past have given us feedback that NuGet does not gracefully handle this scenario and will crash with a very obscure error message that customers don't understand.
+
+This could be mitigated by persisting clear test API keys in a completely different section of the `nuget.config` file, such as `<clearApiKeys>`, which older versions of NuGet should ignore.
+
+2. Writing nuget.config from older versions of NuGet.
+
+If a customer runs any command that requires `nuget.config` to be modified, for example `dotnet nuget add source`, NuGet will deserialize the existing config, modify the value as per the command, and then serialize back to disk.
+Since older versions of NuGet don't know about the `encrypted="true/false"` attribute, this attribute will be lost, and then if a newer version of NuGet is run, it will assume that the value is encrypted, and then cause errors.
+
+If NuGet doesn't already handle maintaining sections of the config file that it doesn't know about, then there is no mitigation for this.
+Using a different section name, like `<clearApiKeys>` will still get lost, but at least it won't cause DPAPI to try to decrypt an unencrypted string, and therefore improving the error message will not be as important.
+
+### Unencrypted secrets
+
+Since the API key is a secret, providing a way to reduce security might not be desirable.
+However, at the time that this design spec is being written, there is no alternative on Linux and Mac, making API keys in the `nuget.config` file a Windows-only feature, which is also not desirable.
```

Comment:
Is the Apple Keyring usable from the command line, or without user interaction? If that is possible then on an MacOS device put the credentials in the keyring, and allow the expanded spec to include something like location in the key syntax, and require either a value or a location.